internal/migrations/changesets.go

```diff
@@ -0,0 +1,319 @@
+	ResolveFunc: func(ctx context.Context, app *common.App, resolution *resolutions.Resolution, block *common.BlockContext) error {
+
+		// Extract the migration config from the app, executed only once
+		if migrationCfg == nil {
+			migrationCfg = &MigrationConfig{}
+			cfg, ok := app.Service.ExtensionConfigs[ListenerName]
+			if !ok {
+				return errors.New("no migration config provided")
+			}
+
+			err := migrationCfg.ExtractConfig(cfg)
+			if err != nil {
+				return err
+			}
+		}
+
+		var migration ChangesetMigration
+		err := migration.UnmarshalBinary(resolution.Body)
+		if err != nil {
+			return err
+		}
+
+		// insert the changeset into the database
+		app.Service.Logger.Debug("insert changeset chunk", log.Int("height", migration.Height.Int64()), log.Int("chunkIndex", migration.ChunkIdx.Int64()))
+		tx, err := app.DB.BeginTx(ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback(ctx)
+
+		// insert the changeset into the database
+		err = migration.insertChangeset(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		// Commit the changeset chunk to the database, so that it is persisted
+		// even if the resolution fails later
+		if err = tx.Commit(ctx); err != nil {
+			return err
+		}
+
+		tx, err = app.DB.BeginTx(ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback(ctx)
+
+		var currentHeight int64
+		// get the last changeset height applied
+		lastChangeset, err := getLastChangeset(ctx, tx)
+		if err != nil {
+			return tx.Commit(ctx)
+		}
+
+		if lastChangeset == migrationCfg.EndHeight {
+			return nil
+		} else if lastChangeset == -1 {
+			currentHeight = migrationCfg.StartHeight
+		} else {
+			currentHeight = lastChangeset + 1
+		}
+
+		// Apply the changesets in order of block height
+		for {
+			// If the current height is greater than the migration end height, break
+			if currentHeight >= migrationCfg.EndHeight {
+				app.Service.Logger.Info("changeset migration completed", log.Int("height", currentHeight))
+				break
+			}
+
+			// Check if all chunks have been received for the current height, if not, break
+			totalChunks, chunksReceived, err := getChangesetMetadata(ctx, tx, currentHeight)
+			if err != nil {
+				break
+			}
+
+			if totalChunks != chunksReceived {
+				// app.Service.Logger.Info("waiting for all chunks to be received", log.Int("height", currentHeight))
+				break
+			}
+
+			// Apply the changeset
+			if err = applyChangeset(ctx, tx, currentHeight, totalChunks); err != nil {
+				return err
+			}
+
+			app.Service.Logger.Info("Applied changesets", log.Int("height", currentHeight))
+
+			// Delete the changeset after it has been applied
+			if err = deleteChangesets(ctx, tx, currentHeight); err != nil {
+				return err
+			}
+
+			// Increment the last changeset
+			if err = setLastChangeset(ctx, tx, currentHeight); err != nil {
+				return err
+			}
+
+			currentHeight += 1 // move to the next height
+		}
+
+		return tx.Commit(ctx)
```

Comment:
I know this is a somewhat substantial change, but after looking at this, I think there is potentially a less error prone way to do it:

Within each changeset chunk's resolution, you should store the data locally as a part of some database that is tied to the networks state (you already do this). Then, you have sn `EndBlockHook` that, while the migration is running, queries the local database for any blocks where it has all of the chunks, but has not applied them, and then applies them and deletes them from the local store.